KAFKA-9441: Unify committing within TaskManager

mjsax opened the pull request:
With KIP-447 we need to commit all tasks at once using eos-beta as all tasks share the same producer. Right now, each task is committed individually (by making multiple calls to `consumer.commitSync` or individual calls to `producer.commitTransaction` for the corresponding task producer).

To allow for a unified commit logic, we move the commit logic into `TaskManager`. For non-eos, we collect all offset to be committed and do a single call to `consumer.commitSync` -- for eos-alpha we still commit each transaction individually (but now triggered by the `TaskManager` to have all code in one place).

To allow for a unified commit logic, we need to split existing method on `Task` interface in pre/post parts to allow committing in between, in particular:

 - `commit()` -> `prepareCommit()` and `postCommit()`
 - `suspend()` -> `prepareSuspend()` and `suspend()` (we keep the name as it still does suspend the task)
 - `closeClean()` -> `prepareCloseClean()` and `closeClean(Map<TopicPartitions, Long> checkpoint)` (we keep the name as it still does close the task)
 - `closeDirty()` -> `prepareCloseDirty()` and `closeDirty()` (we keep the name as it still does close the task)

`prepareCloseClean()` returns checkpoint information to allow checkpointing after the `TaskManager` did the commit within `closeClean()`.

In a follow up PR, we will introduce eso-beta and commit a single transaction over all task using the shared producer.

Call for review @guozhangwang @abbccdda

mjsax committed:
KAFKA-9441: Unify committing within TaskManager

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -275,6 +277,18 @@
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static final String UPGRADE_FROM_25 = "2.5";
+
```
This is not really relevant for this PR, but we need to add it for KIP-447 eventually, thus I just include it in this PR.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java:
```diff
@@ -45,8 +44,6 @@
                      final Serializer<V> valueSerializer,
                      final StreamPartitioner<? super K, ? super V> partitioner);
 
-    void commit(final Map<TopicPartition, OffsetAndMetadata> offsets);
```
We moved this to `TaskManager`

guozhangwang replied:
Please see my other comment above --- I think it is cleaner to just call `foreach(active-task) task.recordCollector.commit` inside the task-manager; and inside RecordCollectorImpl we check that eosEnabled is always true, otherwise illegal-state thrown.

In the next PR where we have the thread-producer, we could then only create a single `recordCollector` object that is shared among all active tasks and wraps the thread-producer, and then the caller `taskManager` code then can just get one active task and call its record-collector's commit function knowing that is sufficient to actually commit for all tasks since everyone is using the same record-collector.

WDYT?

guozhangwang replied:
After syncing offline about this, I think I'm convinced now that moving this logic into TaskManager is better.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -102,8 +103,9 @@ public void completeRestoration() {
 
     @Override
-    public void suspend() {
+    public Map<TopicPartition, OffsetAndMetadata> suspend() {
```
On `suspend()` and `prepareCommit()` we don't commit yet, but return the offsets that need to be committed

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -294,29 +296,24 @@ public void resume() {
-    /**
-     * @throws TaskMigratedException if committing offsets failed (non-EOS)
-     *                               or if the task producer got fenced (EOS)
-     */
```
We don't commit and thus don't throw any longer

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -338,12 +335,14 @@ private void commitState() {
 
-        // we need to preserve the original partitions times before calling commit
-        // because all partition times are reset to -1 during close
+        commitSensor.record(time.nanoseconds() - startNs);
```
Frankly, not sure if this is correct any longer. What do we want to record with this sensor exactly? Flushing can be expensive and we might want to record it as part of committing -- but I am not sure.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:
```diff
@@ -59,9 +61,8 @@
-                           final Producer<byte[], byte[]> producer) {
-        this(logContext, producer, null, null);
+    public StreamsProducer(final Producer<byte[], byte[]> producer) {
+        this(null, producer, null, null);
```
I am not happy with this rewrite (but as I know that John did some changes in this class in another PR, I just did this hack her for now -- needs some cleanup after a rebase)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -481,14 +486,16 @@ int commitAll() {
-                    task.commit();
-                    commits++;
+                    consumedOffsetsAndMetadataPerTask.put(task.id(), task.prepareCommit());
+                    task.markCommitted();
```
We could also do a second loop over all tasks, _after_ calling `commit(..)` below -- not sure if this is ok as-is?

abbccdda replied:
I would prefer a second loop to guarantee a consistent reflection on the task committed state.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:
```diff
@@ -239,49 +231,6 @@ public void shouldPassThroughRecordHeaderToSerializer() {
     }
 
-    @Test
-    public void shouldCommitViaConsumerIfEosDisabled() {
```
Moved both tests to `TaskManagerTest`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:
```diff
@@ -506,103 +445,12 @@ public void shouldThrowStreamsExceptionOnSubsequentCallIfFatalEvenWithContinueEx
     }
 
-    @Test
-    public void shouldThrowTaskMigratedExceptionOnCommitFailed() {
```
move all 4 tests to `TaskManagerTest`

abbccdda replied:
Checkmark for proving the 6 tests are all migrated.

abbccdda commented:
Could we have a summary for this PR?

mjsax commented:
The summary is not longer than the PR title. We make `TaskManager` the class that is responsible for committing (to allow to commit all tasks at once, instead of each task individually)

abbccdda commented:
The title sounds pretty vague to me. The description could at least include what the committing behavior look like under `TaskManager`, what's the motivation, etc as we are already overloading JIRA 9441. In general we could try to be more specific about the changes in the PR description as I could see you are also adding upgrade flags inside this PR. Major side effects should be better to document at first IMHO.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -283,10 +297,26 @@
+     * similar to the {@link #AT_LEAST_ONCE} case.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static final String EXACTLY_ONCE_BETA = "exactly_once_beta";
```
Did you already update the KIP for the new config?

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -707,6 +710,35 @@ public static String getSharedAdminClientId(final String clientId) {
+        return EXACTLY_ONCE_BETA.equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG));
+    }
+
+    public static boolean eosEnabled(final StreamsConfig config) {
```
What's the benefit of building this as a static helper?

mjsax replied:
We will need this later (ie follow up PR) and it reduced code duplication

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -294,29 +295,19 @@ public void resume() {
-                }
-
-                log.info("Committed");
+            case SUSPENDED:
```
We need to allow committing in `SUSPENDED` state now as we first suspend all tasks and than commit. Cf. `TaskManager#handleRevocation()`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -334,16 +325,22 @@ public void commit() {
      *                               or if the task producer got fenced (EOS)
      */
-    private void commitState() {
+    private void commitState(final boolean writeCheckpoint) {
```
MInor improvement: we include writing the checkpoint and the caller can indicate if it should be written or not.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -409,6 +411,10 @@ public void close() {
+            } else {
+                for (final StreamsProducer streamsProducer : taskProducers.values()) {
+                    streamsProducer.close();
+                }
```
This issues was introduced in the PR that introduced `StreamsProducer` -- we forgot to close them. Fixing this on the side.

abbccdda replied:
Sounds good, just mark that depending on John's fix, we probably don't need to handle this.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -138,7 +142,7 @@ void handleCorruption(final Map<TaskId, Set<TopicPartition>> taskWithChangelogs)
             changelogReader.remove(task.changelogPartitions());
 
-            // mark corrupted partitions to not be checkpointed, and then close the task as dirty
+            // mark corrupted partitions to not be checkpointed, and then close the task and revive the task
```
We call `closeClean` below -- just fixing the comment here for now (\cc @guozhangwang)

Note that we don't commit offsets for this case any longer -- previously, committing offsets "might" have been done with `closeClean()` (even if I believe that the task would be marked as "commitNeeded == false"). We don't let the TaskManager commit offsets here, as it should not be required.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -194,7 +198,9 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
-                    taskProducers.remove(task.id());
+                    if (taskProducers.containsKey(task.id())) {
+                        taskProducers.remove(task.id()).close();
+                    }
```
Similar to above: this issue was introduced in the `StreamsProducer` PR. We nee to close the producer when we remove it.

abbccdda replied:
Probably need to change after rebase

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -345,7 +352,9 @@ void handleLostAll() {
-                taskProducers.remove(task.id());
+                if (taskProducers.containsKey(task.id())) {
+                    taskProducers.remove(task.id()).close();
+                }
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -401,9 +410,12 @@ private void cleanupTask(final Task task) {
 
     void shutdown(final boolean clean) {
         final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
-        final Iterator<Task> iterator = tasks.values().iterator();
```
Not sure why we use an iterator here. Simplifying the code with a `for`-loop

mjsax commented on streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java:
```diff
@@ -213,7 +214,9 @@
 
+    private final MockConsumer<byte[], byte[]> consumer;
     private final MockProducer<byte[], byte[]> producer;
+    private final StreamsProducer streamsProducer;
```
We need to commit explicitly in TTD now to mimic the TaskManger. Hence, we need access to the `consumer` and `streamsProducer`

abbccdda replied:
Makes sense to me.

guozhangwang reviewed (commented):
I'm still in the middle of reviewing this PR, just left a meta question here: previously the commit ordering is

1) flush store, flush producer
2) commit offset
3) write checkpoint file

Now it becomes 1/3/2. It means if we have a crash between 2 and 3, in the past we would likely read from an old checkpointed offset and replay the changelogs, which is fine with or without EOS (with EOS we would just read committed data still so it is safe).

But now if we crash between 3 and 2, it means the checkpoint file has been written, but the offsets are not committed yet, which means we could potentially lose some data. Right?

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -504,8 +497,7 @@ public boolean isProcessable(final long wallClockTime) {
      */
-    @SuppressWarnings("unchecked")
-    @Override
+    @SuppressWarnings({"unchecked", "rawtypes"})
```
Why do we start to suppress warnings?

mjsax replied:
We should have done this from the beginning on... (it's just a "side fix")

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -232,17 +232,16 @@ public void completeRestoration() {
      */
     @Override
-    public void suspend() {
+    public Map<TopicPartition, OffsetAndMetadata> suspend() {
```
Add a comment describing the new return statement.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -500,17 +514,43 @@ int maybeCommitActiveTasksPerUserRequested() {
+    private void commit(final Map<TaskId, Map<TopicPartition, OffsetAndMetadata>> offsetsPerTask) {
+        if (eosEnabled) {
+            for (final Map.Entry<TaskId, Map<TopicPartition, OffsetAndMetadata>> taskToCommit : offsetsPerTask.entrySet()) {
+                taskProducers.get(taskToCommit.getKey()).commitTransaction(taskToCommit.getValue());
```
In EOS beta, we should be able to send out a batch commit instead of individual ones?

mjsax replied:
Correct. Unifying the commit logic as done is this PR allows us to do this in a follow up PR that actually enable producer per thread -- the whole purpose of this PR is to prepare/refactor for this.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -684,44 +686,47 @@ public void shouldRespectCommitNeeded() {
+        assertTrue(task.commitNeeded());
+
+        task.markCommitted();
         assertFalse(task.commitNeeded());
```
I don't think we need to test `assertFalse(task.commitNeeded()` as its outcome is controlled by `task.markCommitted`. So we only need to do it once.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -733,9 +738,9 @@ public void shouldCommitConsumerPositionIfRecordQueueIsEmpty() {
+        final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata = task.prepareCommit();
 
-        verify(recordCollector);
+        assertThat(offsetsAndMetadata, equalTo(mkMap(mkEntry(partition1, new OffsetAndMetadata(3L, encodeTimestamp(0L))))));
```
Why do we no longer have the mock verification?

mjsax replied:
We remove `RecordCollector#commit()` method in this PR and thus we remove the expected call to commit at the beginning of this test -- thus, there is nothing to be verified any longer and we don't call `commit()` with `prepareCommit()` any longer.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -441,10 +458,54 @@ public void shouldCommitActiveAndStandbyTasks() {
+        task01.setCommitNeeded();
+        taskManager.tasks().put(taskId01, task01);
+
+        consumer.commitSync(new HashMap<>());
```
Should we do `expectLastCall` here?

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -812,7 +877,79 @@ public void closeClean() {
+            () -> taskManager.commitAll()
+        );
+
+        assertThat(thrown.getMessage(), equalTo("Timed out while committing offsets via consumer"));
```
We should also verify the thrown cause

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -812,7 +877,79 @@ public void closeClean() {
+            () -> taskManager.commitAll()
+        );
+
+        assertThat(thrown.getMessage(), equalTo("Consumer committing offsets failed, indicating the corresponding thread is no longer part of the group; it means all tasks belonging to this thread should be migrated."));
```
Same here, for verifying the thrown cause

guozhangwang replied:
This is a meta comment: since we moved the commit logic out of the tasks into task-manager already, we should add the check that:

1) inside the task manager, if the commit failed with fatal errors, the corresponding follow-up steps (postCommit, suspend, closeClean) should be skipped, and the exception is thrown out of the task-manager to thread
2) if commit failed with fenced errors, follow-up steps are also skipped (tasks state should be un-changed) and the task-migration exception is thrown out of the task-manager.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -500,17 +514,43 @@ int maybeCommitActiveTasksPerUserRequested() {
+            for (final Map.Entry<TaskId, Map<TopicPartition, OffsetAndMetadata>> taskToCommit : offsetsPerTask.entrySet()) {
+                taskProducers.get(taskToCommit.getKey()).commitTransaction(taskToCommit.getValue());
+            }
+        } else {
```
What's the reasoning her for only wrapping the consumer offset commit case here, not for EOS case?

mjsax replied:
You mean exception handling? For the producer all exception handling is done within `StreamsProducer` (note that `threadProducer` above is a `StreamsProducer`, not a `KafkaProducer`)

abbccdda replied:
👍

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java:
```diff
@@ -200,8 +200,7 @@ private KeyValueStoreTestDriver(final StateSerdes<K, V> serdes) {
             new TaskId(0, 0),
-            consumer,
-            new StreamsProducer(logContext, producer),
+            new StreamsProducer(producer),
```
Always feels better for one less parameter :)

abbccdda reviewed (changes requested):
Thanks for the PR, left some comments

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -92,7 +92,6 @@
     private final Sensor closeTaskSensor;
     private final Sensor processLatencySensor;
     private final Sensor punctuateLatencySensor;
-    private final Sensor commitSensor;
```
This is an open question: we don't want to remove this sensor however it was unclear to me how to handle this metric after we split "task committing" into three steps (prepareCommit; taskManager#commit; postCommit).

abbccdda replied:
Should we attempt to add more fine-grained metrics for 3 stages then?

mjsax replied:
Frankly, I have no good idea atm... Also, if we change metrics, we need to update the KIP and it's getting more complicated. If possible, I would prefer to not change any metric, but not sure if it is possible...

guozhangwang replied:
I actually think that we can remove this DEBUG-level per-task commit metrics, since we already have the INFO-level per-thread commit metric and this one does not provide much more additional information?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:
```diff
@@ -58,25 +58,24 @@
         this.producer = Objects.requireNonNull(producer, "producer cannot be null");
         this.applicationId = applicationId;
-        this.eosEnabled = eosEnabled;
+        this.eosEnabled = applicationId != null;
```
Simplification to avoid passing in `eosEnabled` and reducing constructor parameter list -- we just piggy back on the `application.id` that shall be `null` for non-eos.

vvcephei replied:
It seems a bit roundabout to have to remember we should send a `null` `application.id` as the constructor argument to indicate that eos is enabled. What's wrong with saying "eos is enabled" when you want eos to be enabled?

guozhangwang replied:
Subjectively I'd +1 that adding one more parameter to avoid piggy-backing on the applicationId is better.

mjsax replied:
It's a personal preference I guess. But seems you don't like it. Will revert it.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:
```diff
@@ -58,25 +58,24 @@
 
     private String formatException(final String message) {
-        return message + " [" + logPrefix + ", " + (eosEnabled ? "eos" : "alo") + "]";
+        return message + " [" + logPrefix + "]";
```
Avoid redundant logging.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:
```diff
@@ -58,25 +58,24 @@
      * @throws IllegalStateException if EOS is disabled
      */
-    public void initTransaction() {
+    void initTransaction() {
```
Side cleanup: All those method can actually be package private.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java:
```diff
@@ -71,7 +72,6 @@
         RESTORING(2, 3, 4),    // 1
         RUNNING(3, 4),         // 2
         SUSPENDED(1, 4),       // 3
-        CLOSING(4, 5),         // 4, we allow CLOSING to transit to itself to make close idempotent
```
Removing this state -- this is an open question if I did this correctly. \cc @vvcephei

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -385,30 +387,30 @@ public void shouldConstructMetricsWithBuiltInMetricsVersionLatest() {
-            task.id().toString(),
-            builtInMetricsVersion
-        ));
+//        assertNotNull(getMetric(
```
After we addressed the question how we want to do metrics, we can update this tests

mjsax commented on streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java:
```diff
@@ -213,7 +214,9 @@
 
+    private final MockConsumer<byte[], byte[]> consumer;
     private final MockProducer<byte[], byte[]> producer;
+    private final TestDriverProducer testDriverProducer;
```
Because we make app method in `StreamsProducer` package private but need access to `commit()` we add `TestDriverProducer` to get access.

mjsax commented on streams/test-utils/src/main/java/org/apache/kafka/streams/processor/internals/TestDriverProducer.java:
```diff
@@ -0,0 +1,38 @@
+        super(producer, applicationId, logContext);
+    }
+
+    public void commitTransaction(final Map<TopicPartition, OffsetAndMetadata> offsets) throws ProducerFencedException {
```
Just added to give public access to `commitTransaction()` to TTD (it's more elegant than making `StreamsProducer#commitTransaction` public IMHO)

guozhangwang replied:
Why making `commitTransaction` is less elegant? I thought that was fine since `StreamsProducer` is inside the internals package anyways? In fact, in TTD we have access to InternalTopologyBuilder accessing it functions (we used to also have a wrapper of InternalTopologyBuilder which we removed later) so I thought that was the agreed pattern.

mjsax replied:
It's obviously subjective -- personally, even if something is internal, we should not just declare stuff as `public` but try to keep it to a minimum to follow the idea of encapsulation (not always possible). If you want me to remove this class and make the method `public` I can do it in a follow up PR. Not sure if we have an agreed pattern, though.

guozhangwang replied:
Cool, in that sense let's just keep it then -- do not add it in one PR and remove it immediately in the next.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -80,32 +80,42 @@ private static String getTaskProducerClientId(final String threadClientId, final
             threadProducer = null;
             taskProducers = new HashMap<>();
         } else {
+            log.info("Creating thread producer client");
```
nit: we could log thread-id here for easier log search.

mjsax replied:
The `threadId` is already added to the log prefix when the `log` object is created in `StreamsThread`

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -182,30 +193,49 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
                 try {
-                    task.closeClean();
+                    checkpointPerTask.put(task, task.prepareCloseClean());
+                    if (task.state() != CREATED) {
```
Could we internalize this state check inside the task to simplify the logic here?

mjsax replied:
`prepareCloseClean()` already does a state check and returns `emptyMap` if state is `CREATED`.

The point of this check is, that we don't add anything to the `consumedOffsetsAndMetadataPerTask` map -- this is important for the corner case for which all tasks are in state CREATED and thus no transaction was initialized. For this case we cannot call `producer.addOffsetsToTranscation()` and must skip this step entirely. Note, that we have a corresponding check below to not call `commitOffsetsOrTransaction` if the map is empty.

guozhangwang replied:
See my other comments: we should not commit in CREATED, RESTORING and SUSPENDED state, and it's better just to let the prepareXX function to indicate if there's anything to commit based on its state internally than letting task-manager to branch on the task state -- more specifically, here the prepareClose call should not return the map of checkpoints but the map of partition -> partition-timestamps (if empty it means nothing to commit), since the checkpoint map are not needed at task-manager at all and post commit, if the offsets should be empty it would still be empty.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -320,13 +363,27 @@ boolean tryToCompleteRestoration() {
-                task.suspend();
+                task.prepareSuspend();
+                if (task.state() != CREATED) {
+                    consumedOffsetsAndMetadataPerTask.put(task.id(), task.committableOffsetsAndMetadata());
```
Similarly here, this state check could be internalized.

guozhangwang replied:
I think we should just let the prepareXX function to return the map of partitions -> partition-timestamp to indicate if it should be included in the map of committing offsets, so that we do not need to leak the state into task-manager here. Also we only need to call `mainConsumer.position` once for all tasks -- please see my other comment above.


Also: we should not try to commit state if we are in RESTORING but only flushing store and writing checkpoints (I think this is already the behavior in trunk), since the partitions are paused from the main-consumer before restoration is done --- maybe it is partly causing some unit test flakiness.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java:
```diff
@@ -123,35 +123,41 @@ public boolean isValidTransition(final State newState) {
      * @throws StreamsException fatal error, should close the thread
      */
-    void closeClean();
+    Map<TopicPartition, Long> prepareCloseClean();
```
nit: let's order the functions as 
```
prepareCloseClean
closeClean
prepareCloseDirty
closeDirty
```

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java:
```diff
@@ -123,35 +123,41 @@ public boolean isValidTransition(final State newState) {
+    Map<TopicPartition, Long> prepareCloseClean();
 
     /**
      * Close a task that we may not own. Discard any uncommitted progress and close the task.
```
Prepare to uncleanly close a task that we may not own.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -144,6 +149,7 @@ void handleCorruption(final Map<TaskId, Collection<TopicPartition>> taskWithChan
             final Collection<TopicPartition> corruptedPartitions = entry.getValue();
             task.markChangelogAsCorrupted(corruptedPartitions);
 
+            task.prepareCloseDirty();
```
Having a `prepareCloseDirty` makes the calling of `closeDirty` a bit cumbersome as we always need to call `prepareCloseDirty` first. To simplify or just do a reminder, I have two suggestions:

1. Internally create a task state called PREPARE_CLOSE or just a boolean like `closeDirtyPrepared` as the state check, so that closeDirty will throw illegal state if the flag is false
2. Following #1, instead of throw, if we don't see the prepareClose is being called, the `closeDirty` will invoke `prepareCloseDirty` first internally.

mjsax replied:
I actually had a similar though, but was not sure if it's worth it. Would like to hear from @guozhangwang @vvcephei what they think?

If we do this, we might want to do it for "commit" and "suspend", too. For suspend() adding a state SUSPEND_PREPARED is not helpful as suspend() does different things depending on the previous state. (For commit and close an additional state would work). For consistency reasons, an internal flag might be better though.

Not sure ate if calling "prepare" automatically would actually be correct for all cases?

mjsax replied:
We can address this in a follow up PR.

abbccdda replied:
Yea, a TODO is also ok.

guozhangwang replied:
`For the next PR`: I think we can save prepareClose (or more accurately, merge prepareClose and close together again) if we make a state diagram change that only suspended state can transit to closed state, i.e. at task-manager level whenever we want to close a task we call its `suspend` function first, which would, depending on its state, be a no-op, or flushing, or closing topology etc, and then after that the task is always in SUSPENDED state, and then we call "commit" if necessary, and then we call close (a minor thing is that today when the state is in SUSPENDED we would omit committing inside task, and we need to lift this restriction; and also the transition actions to transit to SUSPENDED need to rely on the clean flag, hence we need `suspend(clean-flag)`).

AND we can further merge prepareSuspend and suspend as well by just making the checkpointing logic as part of post-commit instead of post-suspend, since as I mentioned above you only have three cases:

1) do not need to checkpoint: if you are in CREATED.
2) checkpoint written and consumed offsets: if you are in RUNNING, in which you need to commit offsets as well.
3) checkpoint only store offsets: if you are in RESTORING, and in which case you do not need to commit offsets.

In fact, if we are not in the RUNNING state yet, the `consumedOffsets` as well as `recordCollector#offsets()` are always going to be empty, so it is always safe to call `stateMgr.checkpoint(checkpointableOffsets())` and not condition on the state and call `stateMgr.checkpoint(emptySet())`.

And if we now allow committing in SUSPENDED state as part of closing (i.e. suspend -> commit -> close), similar rules apply: if we are suspending from a RESTORING state, then in `postCommit` while we ``stateMgr.checkpoint(checkpointableOffsets())` the `checkpointableOffsets` would always be empty; if we are suspending from a RUNNING state it would contain some offsets.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -555,14 +653,46 @@ int maybeCommitActiveTasksPerUserRequested() {
-            int commits = 0;
+            final Map<TaskId, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadataPerTask = new HashMap<>();
             for (final Task task : activeTaskIterable()) {
                 if (task.commitRequested() && task.commitNeeded()) {
```
@guozhangwang For my own education, why we do `and` here instead of just checking `commitRequested`?

mjsax replied:
To avoid the overhead to commit offset that are already committed, ie, the previous commit committed offset 5 and now we would commit offset 5 again.

guozhangwang replied:
"as above" :)

mjsax replied:
Within `maybeCommitActiveTasksPerUserRequested` we know that we are in state `RUNNING` and thus, no need to check what `committableOffsetsAndMetadata()` returns but we can "blindly" commit.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -115,48 +121,45 @@ public void resume() {
+        }
+    }
 
+    @Override
```
Comment here for no better place: Standby task always returns an empty `committableOffsetsAndMetadata`, then why do we still need to check `commitNeeded` for it? Shouldn't it always set to false?

mjsax replied:
Not sure if I can follow. We don't check `commitNeeded` in `postCommit()`? Can you elaborate?

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -80,32 +80,42 @@ private static String getTaskProducerClientId(final String threadClientId, final
-            log.info("Creating thread producer client");
+
+            final String logPrefix = String.format("stream-thread [%s] ", Thread.currentThread().getName());
+            final LogContext logContext = new LogContext(logPrefix);
```
`logContext ` is not used.

mjsax replied:
Good catch.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -410,54 +410,88 @@ public void closeDirty() {
      *                               or if the task producer got fenced (EOS)
      */
-    private void close(final boolean clean) {
+    private Map<TopicPartition, Long> prepareClose(final boolean clean) {
```
Could we add a `@return` for this method? Also we should comment about the different indications when we return an empty map vs null.

mjsax replied:
Why do we need to document this in the method JavaDoc? It's an internal method? Internal comment outdate quickly if code is changed and comments are not updated accordingly (what happens 99% of the time). Hence, I would prefer to limit comments if possible. In doubt, we should document at `Task` level anyway.

mjsax replied:
@guozhangwang I am actually wondering about point (5) -- why do we need to checkpoint the state manager if we wipe out the store later anyway for the unclean EOS case?

guozhangwang replied:
Yes we are unnecessarily checkpointing here --- the reason is that EOS flag was original striped out of task and only processor-state-manager knows about it; now since we get this EOS flag back to task (sigh.. :) we can add this additional check.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -410,54 +410,88 @@ public void closeDirty() {
         } else if (state() == State.SUSPENDED) {
-            // do not need to commit / checkpoint, since when suspending we've already committed the state
-            transitionTo(State.CLOSING);
+            // if `SUSPENDED` do not need to checkpoint, since when suspending we've already committed the state
```
Remove `if`

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -182,30 +193,49 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+        for (final Map.Entry<Task, Map<TopicPartition, Long>> taskAndCheckpoint : checkpointPerTask.entrySet()) {
+            final Task task = taskAndCheckpoint.getKey();
+            try {
+                task.closeClean(checkpointPerTask.get(task));
```
I feel a bit weird here as we don't need `prepareCloseClean` anymore. This API usage is a little complicated upon when we should do it and we don't.

mjsax replied:
We need to call `prepareCloseClean` (as done in L196 above) _before_ we call `commitOffsetsOrTransaction` (L215 above).

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -182,30 +193,49 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                taskCloseExceptions.put(task.id(), e);
+                // We've already recorded the exception (which is the point of clean).
+                // Now, we should go ahead and complete the close because a half-closed task is no good to anyone.
+                task.closeDirty();
```
Similarly for `closeDirty` and `prepareCloseDirty`

mjsax replied:
Some comment as above.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -222,7 +225,8 @@ public void shouldDoNothingWithCreatedStateOnClose() {
 
         task = createStandbyTask();
-        task.closeClean();
+        final Map<TopicPartition, Long> checkpoint = task.prepareCloseClean();
```
nit; 228 - 229 could be merged.

mjsax replied:
I think it's easier to read if it's split.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -410,54 +410,88 @@ public void closeDirty() {
      *                               or if the task producer got fenced (EOS)
      */
-    private void close(final boolean clean) {
+    private Map<TopicPartition, Long> prepareClose(final boolean clean) {
```
Also the above step #4 is no longer correct, the commit is done on TaskManager now.

mjsax replied:
You see -- that is may point from above... The code should be written in a way that explains itself... Updating comments always slips...

abbccdda replied:
I couldn't fully follow this idea, just playing devil advocates here, if we think meta code comments actually hinder the readability of internal class, why not just remove all the internal function meta comments, as they would get outdated anyway? For me the return type comment is still valuable for understandability. If the comment gets outdated, we should just update it. cc @guozhangwang if the idea here makes sense, or we could get a consensus on what needs to be done in internal class comments, and what's not.

guozhangwang replied:
I would suggest not restricting ourselves to some specific rules about comments :) Personally I tried to avoid the `one line comment explaining one line code` type of comments inside a function since it should be obvious, rather I'd add some comments for a block or several blocks if I fear it maybe hard to read by itself. I think you guys should just make your best judgement here.

And for internal functions, I agree that we do not necessarily need to write java-docs, and this one, for example, I wrote the java-doc as part of the tech debt cleanup just to remind what operations MUST be considered here inside closing / suspending etc so that later on when we change the function itself by other contributors, they would use it as a reference to check if they mistakenly missed some steps or re-ordered some steps. However if we are going to split this function into multiple, instead of just re-structuring the function as a whole, then although I have my preference I'd leave to you guys if you want to add the javadoc for both pre/post of you feel now it is too obvious to bother :)

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -432,24 +490,54 @@ private void cleanupTask(final Task task) {
             }
+        }
+
+        for (final Task task : dirtyTasks) {
```
Do we need to keep a task once it is failed to clean close? Why couldn't we just close it dirty immediately after we see the exception?

mjsax replied:
Good idea!

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java:
```diff
@@ -88,44 +88,11 @@ public void before() {
         eosStreamsProducer.initTransaction();
     }
 
-    @Test
```
Why do we need to move these tests?

mjsax replied:
I try to keep "order" and group test methods to keep an overview if test coverage is complete.
```
// generic tests
    // functional
    // exception handling
// non-EOS tests
    // functional
    // exception handling
// EOS tests
    // functional
    // exception handling
```

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -364,11 +365,8 @@ private void commitState() {
-        commitNeeded = false;
-        commitRequested = false;
-        commitSensor.record(time.nanoseconds() - startNs);
+        return consumedOffsetsAndMetadata;
```
Looks like we lack test coverage for TimeoutException and KafkaException cases

mjsax replied:
Yeah, this PR does not yet add all required test...

mjsax replied:
Covered via `shouldCommitNextOffsetFromQueueIfAvailable` and `shouldCommitConsumerPositionIfRecordQueueIsEmpty`

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -247,6 +277,19 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+        if (task.isActive()) {
+            try {
+                activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(task.id());
+            } catch (final RuntimeException e) {
```
We don't have unit test coverage for this exception case

mjsax replied:
Added test `shouldThrowWhenHandlingClosingTasksOnProducerCloseError`

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -182,30 +193,49 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+            final Task task = taskAndCheckpoint.getKey();
+            try {
+                task.closeClean(checkpointPerTask.get(task));
+            } catch (final RuntimeException e) {
```
We lack unit test coverage for this case

mjsax replied:
I know...

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -212,23 +227,24 @@ public void shouldCloseActiveTasksWhenHandlingLostTasks() {
         final Task task00 = new StateMachineTask(taskId00, taskId00Partitions, true);
         final Task task01 = new StateMachineTask(taskId01, taskId01Partitions, false);
 
+        // `handleAssignment`
```
Could we verify the assignment stack and lost stack separately, by doing `handleAssignment` verify first before calling `handleLost`

mjsax replied:
Not sure if I can follow? The comments just mark which setup calls belongs to which test call, nothing more. All setup is done upfront before we call the actually methods under test.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -80,32 +80,42 @@ private static String getTaskProducerClientId(final String threadClientId, final
+        if (threadProducer != null) {
+            throw new IllegalStateException("Producer per thread is used");
+        }
+        return taskProducers.get(taskId);
```
nit: add a check that taskId exists in `taskProducers` to make sure we do not return null.

guozhangwang replied:
Actually on a second thought, I'm wondering if the following inside TaskManager is cleaner:

```
for (task <- taskManager.activeTasks)
    task.recordCollector().commit(taskToCommit.getTask(task.id);
```

Instead of:

```
activeTaskCreator.streamsProducerForTask(taskToCommit.getKey()).commitTransaction(taskToCommit.getValue());
```

My gut feeling is that it is cleaner to not access the task creator for its created stream-producers (and hence here we need to change the task-producer map to streamsProducers), but just access each task's record collector and call its `commit` --- today we already have a `StreamTask#recordCollector` method.

mjsax replied:
I think it's unclean to let the RecordCollector commit (note that this PR removes `RecordCollector` not at side refactoring but on purpose) -- to me the RecordCollector has the responsibility to bridge the gap between the runtime code (that is typed), and the Producer that uses `<byte[],byte[]>` (ie, it serialized the data and manages errors from `send`) -- why would a **_collector_** know anything about committing (for which it also needs a handle to the consumer)?

About accessing the `ActiveTaskCreator`: we could also expose the `StreamsProducer` via the `RecordCollector` though (or directly via the task)? That would be cleaner I guess.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -115,48 +121,45 @@ public void resume() {
-
-            default:
-                throw new IllegalStateException("Illegal state " + state() + " while committing standby task " + id);
+    public void prepareCommit() {
```
This is a meta comment: I think we can consolidate `prepareCommit` and `prepareClose` and `prepareSuspend` here by introducing the clean parameters to the function, since their logic are very similar (for the part that they diffs a bit, it can be pushed to post logic), and on task-manager during commit:

1) for each task -> task.prepareCommit(true)
2) commit
3) for each task -> task.postCommit(true)

During close:

if (clean)
    1) for each task -> task.prepareCommit(true)
    2) commit()
    3) for each task -> task.postCommit(true)
else
    1) for each task -> task.prepareCommit(false)
     // do not commit
    3) for each task -> task.postCommit(false)
4) tasks.close(flag)

And the same for suspension.

mjsax replied:
Will do this in a follow up PR.

guozhangwang replied:
SG.

I think in this PR we still can do the change to let `prepareXX` to return the map of partitions -> partition-timestamp to indicate whether this task should be included in committing.

guozhangwang reviewed (commented):
This is also a meta comment: I'm trying to think of a way where we do not need to mess the hierarchy of task / task-manager classes which are just cleaned.

One wild thought following my previous comment, with the goal to avoid all of those `prepare` `post` ordering etc, is to abstract this logic away from the task / task-manager and handle inside the record-collector. In other words, the task still follow the same pattern as in today's trunk, e.g. in `commitState`, calling `recordCollector.commit(consumedOffsetsAndMetadata);`. However, the recordCollector does not call the corresponding commitTxn immediately but wait until all expected `commit` calls as been triggered and then call the function with the aggregated offset map. More specifically:

1. We let the txn-manager to keep a reference to the shared record-collector.
2. We add a "pre-commit" function inside record-collector which passed in the set of expected tasks (or partitions?) to be committed; and then when `commit` is called, if there is no set of expected tasks set, record-collector would trigger immediately, otherwise, wait until all the expected elements have been passed in, and then trigger.
3. In these scenarios:

3.a) Suspend: Although we may only suspend a subset of tasks, we'd still have to commit ALL tasks under eos-beta, so we just call `record-collector.preCommit` with all the tasks, and then forloop task.suspend.

3.b) Commit: we would have to commit all tasks, so we just call `record-collector.preCommit` with all the tasks, and then forloop task.commit.

3.c) CloseClean: no matter what task(s) we are closing, we need to committing for all, so the same as above.

3.d) CloseDirty: we do not need to commit at all, so we do not need to call `record-collector.preCommit` since we know its `commit` function would not be triggered.

I admit it is not ideal since we are sort of poking a hole inside record-collector to be tasks-aware, but it saves all the code changes we'd have to introduce in task and most of the task-manager messiness.


@mjsax sorry for going back and forth on the high-level code design here, I know changing 1000+ LOC is not a easy job... but I just want to make sure we introduce as less tech complexity in our code base as possible to achieve the same thing.

guozhangwang commented:
Thanks for the rebasing @mjsax ! will take another look soon.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -293,18 +302,31 @@ public void resume() {
             case RESTORING:
-                commitState();
+                commitNeeded = false;
+                commitRequested = false;
```
This is not introduced in this PR, but: while thinking about it, I realized for RESTORING state we do not need to rely on eosDisabled to checkpoint, in fact we can always checkpoint during RESTORING here.

mjsax replied:
Ack

abbccdda replied:
Do we have unit test to check the checkpoint status after `postCommit()`?

mjsax replied:
Yes, `shouldRespectCommitNeeded()` check this already.

guozhangwang reviewed (commented):
I made another look at the PR, and here's another proposal for reducing the added Task APIs:

1) We still make all the state-transition functions plus the commit call to be state-dependent, and not leak to the task-manager. More specifically:

1.a) prepareCommit: should include `stateMgr.flush()` as long as state is not CREATED / SUSPENDED; should include `recordCollector.flush()` if state is RUNNING. The function should return the map of offsets that should be committed. If pre-commit is a no-op based on the state, then it returns an empty map.

1.b) postCommit: if state is RUNNING then we should `stateMgr.checkpoint(checkpointableOffsets())` if `eosDisabled`; if it is RESTORING or CREATED or we always blindly `stateMgr.checkpoint(emptyMap())`; in other states we do nothing.

1.c) commit: take the aggregated map from the pre-commit on tasks, make the call on consumer or producer, and capture the exceptions and handle accordingly (see below).

2. We extract the decision of whether we should commit (including pre and post) during suspend / close out of the task, into task-manager. In fact, we always try to commit ALL tasks when suspending / closing ANY tasks. 

2.a) For suspending, inside task-manager we first trigger “suspend” for those tasks taht are decided for suspension. The function would close topology / clear partition-group depending on the state of the task. And then we trigger the commit procedure for all tasks.

According to 1) above, for those already-suspended / created tasks, pre-commit would be no-op and return nothing, and hence effectively they would not be included for committing; if everyone returns empty in pre-commit, i.e. the aggregated map is also empty, we can skip the committing call, BUT we do not skip the post-commit since even if we do not commit offsets, we may still want to write checkpoint files. If commit fails, we would skip the `post-commit` call and the thrown exception would be handled at the thread-level (either handleable task-migrated, or fatal streams exception).

With that we can save pre-suspend.

2.b) For closing, it is the most tricky one, as we need to depend on the “clean” flag. The key point is that, post-447 if one task is being closed dirty then we’d have to abort the whole txn of the shared thread-producer to make sure the downstream would not read any records of this dirty-closed task (that’s the guaranteed pre-447, which we need to maintain). This is a tradeoff we have to make in 447 but hopefully the close-dirty case is rare and when we close-dirty, we always close-dirty EVERYONE anyways:

* When the thread shutdown-dirty due to exception causing it to die
* When we handleLostAll

And:


* When close-clean throws, instead of capturing it and immediately close-dirty, post-447 we should just let it to be thrown all the way to thread and let thread shutdown all tasks as dirty, since when it happens, none of the records sent by this producer should be committed.
* When handleCorruption, which is ONLY triggered during restoration, we actually do not need to call close-dirty (we originally call it only to trigger the wipe-Store function), instead we just hard-code the close-wipe-and-revive logic inside the task manager itself.

As a result, when we close dirty, we always close dirty all tasks anyways. So in this case we do not need to commit (including the pre-commit which flushes state stores and record collectors) and just trigger `close(dirty)` on each task which would, depending on the state, call `closeTopology`, `closeStateManager` (which would flush the stores while closing), `recordCollector.close`, `partitionGroup.close`, `closeSensor.record`, `wipeStore-if-EOS` respectively and swallow any exceptions thrown. Then if it is eos-beta we let the producer to abort the txn; if it is eos-alpha only the corresponding task-producers would abort txn and close.

On the other hand, if we are going to close-clean any task(s), then we still need to commit all tasks. we would trigger a `suspend` first which would, depending on the state, call `closeTopology` or do nothing, and then we proceed with the commit procedure -- and if it throws, we skip the rest and just let the thread handles it by closing dirty -- and finally we call `close(clean)` and if it throws then we still let thread handles it.

With that we can save pre-close.

2.c) When committing, again we always commit all tasks, and we just follow the pre/commit/post procedure here, which would execute correspondingly based on the states (for some tasks it would be no-op, again).

2.d) Resuming is a dummy transition so nothing changes.

mjsax commented:
Rebased to resolve merge conflicts. Addressed more review comments and added more tests.

Also added a fix (from https://github.com/apache/kafka/pull/8301) as `trunk` is broken atm, to make Jenkins pass.

Triggered systems tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3833/

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -80,32 +80,44 @@ private static String getTaskProducerClientId(final String threadClientId, final
         }
+    }
 
+    StreamsProducer streamsProducerForTask(final TaskId taskId) {
```
We need a unit test for this function.

guozhangwang replied:
`For the next PR` (all other comments with this tag means no changes required for this PR): my understanding is that we would make the thread-producer also a `StreamsProducer` instead of a `KafkaProducer` which would be used to `commitTransaction` under `eosBeta`, is that right?

mjsax replied:
Correct. For eos-beta there will be one `StreamsProducer` that is shared over all tasks.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -231,27 +228,40 @@ public void completeRestoration() {
+            recordCollector.flush();
 
-            // we should also clear any buffered records of a task when suspending it
+            log.info("Prepare suspending running");
```
we could just do one log in front: `log.info("Prepare suspending {}", state());`

mjsax replied:
Well, we log "skip" for state CREATED and we throw for invalid states. Note sure how to do this?

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -400,65 +422,88 @@ public void closeDirty() {
+                // if the latter throws and we re-close dirty which would close the state manager again.
+                executeAndMaybeSwallow(
+                    clean,
+                    () -> {
```
nit: { could be reduced.

abbccdda reviewed (commented):
Thanks for the update, just a couple of minor comments. Overall I think we are good

guozhangwang reviewed (commented):
@ableegoldman mentioned some of the system test failures are consistent in trunk, just FYI.

ableegoldman commented:
I linked the system test failures I saw on the meeting notes for this week, but just to clarify:
1) Two failing consistently: `StreamsEOSTest.test_failure_and_recovery{_complex}` 
2) One flaky: `StreamsStandbyTask.test_standby_tasks_rebalance`
Anything outside of that is a new failure

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -178,18 +194,18 @@ void closeThreadProducerIfNeeded() {
+        if (taskProducer != null) {
             try {
-                producer.close();
+                taskProducer.kafkaProducer().close();
```
nit: we can have a wrapped StreamsProducer#close / metrics, and then #kafkaProducer would be for testing-only.

mjsax replied:
We could, but the idea was that `ActiveTaskCreator` creates the producer via `new KafkaProducer()` and thus it should call `KafkaProducer#close()`, too, and not delegate it to `StreamsProducer`. Thoughts?

guozhangwang replied:
Hmm, I think moving forward we would create and maintain both the single thread-producer and task-producers as StreamsProducer objects right?

mjsax replied:
SGTM

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -313,35 +334,23 @@ public void commit() {
-        // we need to preserve the original partitions times before calling commit
-        // because all partition times are reset to -1 during close
+    @Override
+    public Map<TopicPartition, OffsetAndMetadata> committableOffsetsAndMetadata() {
```
In either eos-alpha or eos-beta or non-eos, we can just loop over all the "committable partitions" and call `mainConsumer#position` once, so this function can be extracted out of the task as a per-task call.

More specifically, in the prepareXX calls, we know based on the state of the task and clean flag whether or not we should commit the source topic offsets for this task, so we can let the prepareXX function to return `Map<TopicPartition, Long> partitionTimes` encoding the extracted timestamps for each partition instead of void --- when we decided not to commit we return an empty map. And then inside TaskManager we just use the `mainConsumer` to call position once and then pass that to the `commitOffsetsOrTransaction` call.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -166,29 +169,70 @@ public void closeDirty() {
+                       final Map<TopicPartition, Long> checkpoint) {
+        if (state() == State.CREATED) {
+            // the task is created and not initialized, do nothing
+            closeTaskSensor.record();
```
nit: we can do `if / else if / else` here still and move the `closeTaskSensor.record(); / transitionTo(State.CLOSED);` to avoid duplication.

mjsax replied:
Well, we can, but we get an empty "than block" what is weird:
```
if (state() == State.CREATED) {
  // empty
} else if {
...
}
```

guozhangwang replied:
I'd say we can always log a debug there saying "doing nothing in this function since we are in this state" :) The main concern I had, is that if in the future we want to add more steps in addition to `recording sensor` etc, we may forget adding it in one place or the other. Removing duplicated code helps us to be less vulnerable.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -166,29 +169,70 @@ public void closeDirty() {
-                    commit();
-                }
+            return;
+        }
```
Ditto here, I think if / else if / else is more readable.

mjsax commented:
Thanks for pointing out. In the run I triggered, the following failed:
 - `StreamsCooperativeRebalanceUpgradeTest.test_upgrade_to_cooperative_rebalance`
 - `StreamsUpgradeTest.test_metadata_upgrade`
 - `StreamsEosTest.test_failure_and_recovery`
 - `StreamsEosTest.test_failure_and_recovery_complex`
 - `StreamsEosTest.test_rebalance_complex`
 - `StreamsEosTest.test_rebalance_simple`

Not sure if this is any helpful to have a safety net for this PR...

mjsax committed:
Github comments

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -231,27 +228,40 @@ public void completeRestoration() {
     @Override
-    public void suspend() {
-        if (state() == State.CREATED || state() == State.CLOSING || state() == State.SUSPENDED) {
+    public void prepareSuspend() {
```
`For the next PR`: as I mentioned in the last commit I feel `prepareSuspend` and `prepareClose` can be consolidated with `prepareCommit` but in the next PR these logic would be changed again for eos-beta so maybe we cannot do that any more, so I'm fine with keeping as-is and we can revisit to see if we can really do this refactoring or not in the next PR when we did the eos-beta.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -400,65 +422,88 @@ public void closeDirty() {
+     * @throws TaskMigratedException if the task producer got fenced (EOS)
      */
-    private void close(final boolean clean) {
+    private Map<TopicPartition, Long> prepareClose(final boolean clean) {
```
`For the next PR`: I see the reason I return the checkpoint is that we are now extracting the committing out of the task and I need to remember if we need to checkpoint and if yes which offsets after we've flushed and before we checkpoint, but since the state of the task would not change before / after the commit during close. 

More specifically we only have three cases: 1) to not write checkpoint, 2) write checkpoints for written offsets (changelogs) only, 3) write checkpoint for written and consumed offsets. And no matter which case it is during the `preClose`, it would always be the same in the `post`, so why do we need to return it to task-manager, book-keep there, and then after commit to pass it back to tasks?

abbccdda replied:
By `next PR`, you mean the one after we finish the EOS-beta commit feature right?

guozhangwang replied:
I mean the next PR when we add the EOS-beta feature --- please see the first comment I have with this tag.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java:
```diff
@@ -125,35 +125,41 @@ public boolean isValidTransition(final State newState) {
 
     /**
-     * Close a task that we still own. Commit all progress and close the task gracefully.
+     * Prepare to close a task that we still own and prepare it for committing
```
nit: we should emphasize that PrepareClose and close calls should be implemented idempotent since we may call it multiple times if a task close clean first and then fail and then close dirty.

mjsax commented:
Address review comments and rebased to pick up fixed from `trunk`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -497,24 +555,49 @@ private void cleanupTask(final Task task) {
                 try {
-                    task.closeClean();
+                    checkpointPerTask.put(task, task.prepareCloseClean());
+                    if (task.state() != CREATED) {
```
Same here: not only CREATED, but also RESTORING and SUSPENDED tasks should not be included in `consumedOffsetsAndMetadataPerTask` and we should not let the task-manager to peek its state.

guozhangwang reviewed (commented):
I've made a thorough pass over the non-testing code now. There are a couple of places regarding the committing which maybe the reason that system tests fails / unit tests flaky.

mjsax committed:
Github comments

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java:
```diff
@@ -61,7 +61,6 @@
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 
-@SuppressWarnings("unchecked")
```
Thanks for the cleanup!

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -264,7 +265,10 @@ public void shouldAttemptToDeleteStateDirectoryWhenCloseDirtyAndEosEnabled() thr
-        task.transitionTo(Task.State.CLOSING);
+        task.transitionTo(Task.State.RESTORING);
+        task.transitionTo(Task.State.RUNNING);
+        task.transitionTo(Task.State.SUSPENDED);
```
Why we have to transit to SUSPENDED before prepare-closing? Originally we want to check that CREATED state can still trigger close.

mjsax replied:
Ack. We can remove this.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1292,7 +1275,14 @@ public void shouldThrowStreamsExceptionWhenFetchCommittedFailed() {
+    }
+
+    @Test
+    public void shouldThrowIfPostCommittingOnIllegalState() {
```
Not introduced in this PR: could we add test checking CLOSED state should not commit as well?

Also checking SUSPENDED state close-call is no-op.

mjsax replied:
Ack

mjsax replied:
> Also checking SUSPENDED state close-call is no-op.

But in `close()` if state is suspended we might still wipe out the state store -- it's not a no-op

guozhangwang reviewed (commented):
Made a pass on the testing part.

mjsax committed:
Fix bug

mjsax committed:
Github comments

mjsax commented:
Addressed review comments and fix bug exposed by system tests. Also added more tests.

New system test run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3834/

guozhangwang commented:
LGTM. Please feel free to merge after green builds.

abbccdda commented:
The failed test seems to be just flaky
org.apache.kafka.streams.processor.internals.StateDirectoryTest.shouldReturnEmptyArrayIfListFilesReturnsNull

mjsax commented:
Actually the same test failed in both previous runs -- but it passes locally for me. As this PR has nothing to do with the failing test, I might just want go ahead and merge. The last PR that changed the failing test was merged recently: https://github.com/apache/kafka/commit/605d55dc173156471fb05c40d715ab318ce74952

There is one concerning comment on that  PR: `Merged to trunk and 2.5 after tests green locally.` ?

Seems the nightly Jenkins job also fails on this test: - 
 - `trunk` -> https://builds.apache.org/blue/organizations/jenkins/kafka-trunk-jdk11/detail/kafka-trunk-jdk11/1252/tests
 - `2.5` -> https://builds.apache.org/blue/organizations/jenkins/kafka-2.5-jdk8/detail/kafka-2.5-jdk8/65/tests

Did this PR break the Jenkins job? \cc @guozhangwang

abbccdda commented:
@mjsax My understanding is that this PR will fix the problem https://github.com/apache/kafka/pull/8310/

You could choose to merge it first if looks good, and rebase the current one

guozhangwang commented:
@mjsax I just double checked and the failure in trunk is fixed as https://github.com/apache/kafka/pull/8310 while the failure (a different one) in 2.5 is also fixed by @abbccdda 's PR. We should be fine now.

guozhangwang commented:
test this please

mjsax commented:
System test failure `StreamsEosTest` (4 different test as mentioned by Sophie above).

Given that some more fixed got pushed to `trunk` I rebased. This should give us a green Jenkins and we should be able to merge.

ableegoldman commented:
@mjsax only two `StreamsEOSTest` system tests were failing when I ran them. `test_rebalance_complex` and `test_rebalance_simple` are either newly failing or due to this PR

ableegoldman commented:
Also, the two already-failing tests (`test_failure_and_recovery`) failed with a different error than they were trunk:

```
java.lang.IllegalStateException: Unknown TaskId: 0_0
	at org.apache.kafka.streams.processor.internals.ActiveTaskCreator.streamsProducerForTask(ActiveTaskCreator.java:118)
	at org.apache.kafka.streams.processor.internals.TaskManager.commitOffsetsOrTransaction(TaskManager.java:749)
	at org.apache.kafka.streams.processor.internals.TaskManager.commitAll(TaskManager.java:706)
	at org.apache.kafka.streams.processor.internals.StreamThread.maybeCommit(StreamThread.java:772)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:645)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:501)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:475)
```

mjsax committed:
Bug fix

mjsax commented:
@ableegoldman Thanks for pointing it out! Pushed a fix and updated the unit test accordingly. \cc @guozhangwang 

New system test run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3835/

mjsax commented:
`StreamsEosTest.test_failure_and_recovery` and `StreamsEosTest.test_failure_and_recovery_complex` failed. I only see:

```
[2020-03-19 12:12:47,231] ERROR stream-thread [EosTest-e1bb7363-be65-43b7-a35b-326c73533bdf-StreamThread-1] Encountered the following exception during processing and the thread is going to shut down:  (org.apache.kafka.streams.processor.internals.StreamThread)
org.apache.kafka.streams.errors.ProcessorStateException: Error opening store KSTREAM-AGGREGATE-STATE-STORE-0000000003 at location /mnt/streams/EosTest/0_0/rocksdb/KSTREAM-AGGREGATE-STATE-STORE-0000000003
	at org.apache.kafka.streams.state.internals.RocksDBTimestampedStore.openRocksDB(RocksDBTimestampedStore.java:87)
	at org.apache.kafka.streams.state.internals.RocksDBStore.openDB(RocksDBStore.java:195)
	at org.apache.kafka.streams.state.internals.RocksDBStore.init(RocksDBStore.java:231)
	at org.apache.kafka.streams.state.internals.WrappedStateStore.init(WrappedStateStore.java:48)
	at org.apache.kafka.streams.state.internals.ChangeLoggingKeyValueBytesStore.init(ChangeLoggingKeyValueBytesStore.java:44)
	at org.apache.kafka.streams.state.internals.WrappedStateStore.init(WrappedStateStore.java:48)
	at org.apache.kafka.streams.state.internals.MeteredKeyValueStore.lambda$init$0(MeteredKeyValueStore.java:101)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:806)
	at org.apache.kafka.streams.state.internals.MeteredKeyValueStore.init(MeteredKeyValueStore.java:101)
	at org.apache.kafka.streams.processor.internals.StateManagerUtil.registerStateStores(StateManagerUtil.java:81)
	at org.apache.kafka.streams.processor.internals.StandbyTask.initializeIfNeeded(StandbyTask.java:87)
	at org.apache.kafka.streams.processor.internals.TaskManager.tryToCompleteRestoration(TaskManager.java:331)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:586)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:501)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:475)
Caused by: org.rocksdb.RocksDBException: lock : /mnt/streams/EosTest/0_0/rocksdb/KSTREAM-AGGREGATE-STATE-STORE-0000000003/LOCK: No locks available
	at org.rocksdb.RocksDB.open(Native Method)
	at org.rocksdb.RocksDB.open(RocksDB.java:286)
	at org.apache.kafka.streams.state.internals.RocksDBTimestampedStore.openRocksDB(RocksDBTimestampedStore.java:75)
	... 14 more
```
followed by
```
java.lang.NullPointerException
	at org.apache.kafka.streams.processor.internals.StoreChangelogReader.remove(StoreChangelogReader.java:801)
	at org.apache.kafka.streams.processor.internals.TaskManager.cleanupTask(TaskManager.java:549)
	at org.apache.kafka.streams.processor.internals.TaskManager.shutdown(TaskManager.java:567)
	at org.apache.kafka.streams.processor.internals.StreamThread.completeShutdown(StreamThread.java:839)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:484)
```

Seems to be unrelated to this PR. Are we good to merge?

abbccdda commented:
@mjsax We should be good. I have another PR to fix those https://github.com/apache/kafka/pull/8307
